### DISCUSSION TITLE
I2C component detection

### DISCUSSION THREAD

**Discussion by HelmiBeh** [arendst/Tasmota#22166] [Q&A] [answered]
### PROBLEM DESCRIPTION
_A clear and concise description of what the problem is._
Having upgrated my Vindriktning device with some additional sensors, I have an issue with the autodetection of an I2C device, the SGP30. Built a special 14.2.0.4 version containing drivers for SHT3x, IAQ, ENS16x, AHT2x, SGP30 and vindriktning. SHT3x, IAQ, vindriktning, ENS160 and AHT21 will be recognized all time. But SGP30 is only detected, if the module with ENS160 and AHT21 is not in the I2C chain.

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x ] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ x] Device used (e.g., Sonoff Basic): _ESP-12____
- [ ] Tasmota binary firmware version number used: 14.2.0.4_____
  - [ ] Pre-compiled
  - [ x] Self-compiled
- [x ] Flashing tools used: _OTA____
- [x ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
18:36:45.888 MQT: Luft/RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
18:36:46.131 MQT: Luft/RESULT = {"Module":{"18":"Generic"}}
18:36:46.337 MQT: Luft/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"0":"None"},"GPIO3":{"608":"I2C SCL"},"GPIO4":{"640":"I2C SDA"},"GPIO5":{"7456":"VINDRIKTNING"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"0":"None"},"GPIO13":{"0":"None"},"GPIO14":{"0":"None"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

Console output without detection of SGP30:
00:00:00.063 Project tasmota - Tasmota Version 14.2.0.4(tasmota)-2_7_7(2024-09-18T18:20:14)
00:00:00.119 I2C: SHT3X found at 0x44
00:00:00.153 I2C: AHT2X found at 0x38
00:00:00.156 I2C: IAQ found at 0x5a
00:00:00.303 I2C: EN16X found at 0x53

Device detected:
00:00:00.063 Project tasmota - Tasmota Version 14.2.0.4(tasmota)-2_7_7(2024-09-18T18:20:14)
00:00:00.119 I2C: SHT3X found at 0x44
00:00:00.154 I2C: SGP30 found at 0x58
00:00:00.157 I2C: IAQ found at 0x5a

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by Jason2866** [comment]
Most AHT sensors are not following the i2c standard. Not something which can be solved in software.
If you want to use more sensors on the i2c bus do not use any AHT sensor.

**Comment by HofeBY** [comment, marked as answer]
@HelmiBeh
what´s the command output with connected devices:
I2Cdriver
I2Cscan
a Picture of I2C cabeling type and lenght and the used devices. termination ?
your compiling configuration file.
describe your compilation environment. OS, version of VSC, etc.

**Comment by HelmiBeh** [comment]
what´s the command output with connected devices:
I2Cdriver

Luft/RESULT = {"I2CDriver":"8,15,18,43,46,84"}

I2Cscan

Luft/RESULT = {"I2CScan":"Device(s) found at 0x38 0x44 0x53 0x58 0x5a"}

a Picture of I2C cabeling type and lenght and the used devices.

abt 15cm 0.08mm² wire. ESP-12 with boards: GY-SGP30 (https://www.ebay.de/itm/164311829709), SHT3x (https://www.ebay.de/itm/334687210686), IAQ-Core (https://de.aliexpress.com/item/1005004161607094), ENS160+AHT21 (https://www.ebay.de/itm/165689572415)

termination ?

4* 10k in parallel = 2k5 @ 3V3

your compiling configuration file.

[My_ini.zip](https://github.com/user-attachments/files/17083803/My_ini.zip)

describe your compilation environment. OS, version of VSC, etc.

VSC:
Version: 1.93.1 (user setup)
Commit: 38c31bc77e0dd6ae88a4e9cc93428cc27a56ba40
Date: 2024-09-11T17:20:05.685Z
Electron: 30.4.0
ElectronBuildId: 10073054
Chromium: 124.0.6367.243
Node.js: 20.15.1
V8: 12.4.254.20-electron.0
OS: Windows_NT x64 10.0.19045

**Reply by sfromis** [reply]
If adding the module with the AHT21 breaks I2C, you should read the response from Jason2866. They are not "good citizens" on the bus.

BTW. If CO₂ detection is your reason for getting ENS160, I can't recommend that one either. It has a small glob of corroded tin changing resistance with "something" in the air. While that "something" includes CO₂, it cannot tell what's in the air, as numbers depend on several factors.

**Reply by HofeBY** [reply]
@HelmiBeh
I2C devices seem to be OK. You have thrown away all unnessesary devices. I like that !

I asked about the PullUP Terminator of I2C-Bus SDA and SCL.
If you google for every I2C device like "datasheet pdf SHT30" and you look into every datasheet you will see
that the PullUp Resistor for each line SDA&SCL is normal 4,7kOhm.

The question is now if no external Restistor is added, if the ESP-Chip internal activated PullUP on a exaple ESP32C6 is sufficent ?
The answer is no, because on ESP32C6 datasheet is talking about a "week PullUp&PullDown" Resistor and the value is 45kOhm !
Now it´s clear, that several I2C devices are not able to function regular with only ESP internal activated PullUP.

your solution:
If you add 2* 10kOhm in parallel = 5kOhm @ 3V3 at the 1*at_begining & 1*at_theEnd of the I2C Bus lines of each SDA&SCL line this is OK.

If AHT21 is realy a I2C Bus problem, i think it is good to set him on a separete I2C Bus able on ESP32(x) devices.
And then you also have the possibility to store messurements locally on ESP32(x) flash file system.
OTA is also much better.
ESP32C6 was chosen as example, it has hardware Matter included & maybe in future Tasmota will support it.

**Reply by HelmiBeh** [reply]
It didn't seem to be a pull-up problem, as I use normally 4k7 @5V or 2k7 @3V3. My I2C chain with IAQ, SGP30 and SHT31 shows good results. And I will use a second ESP-02 with the same FW and ENS160 and AHT21 as the other I2C chain, which work also well.

**Reply by arendst** [reply]
You might want to try latest dev first as it supports a second i2c bus on esp8266 wit a small selection of drivers (like esp32 does). See i2c md document

**Reply by sfromis** [reply]
However, in this case the "culprit" (AHT21) is not among the sensor drivers supporting bus 2, and as SGP30 is also only with bus 1, separating these does not sound doable?

**Reply by arendst** [reply]
indeed until someone adds second bus support to either of these.

**Reply by HofeBY** [reply]
@HelmiBeh
Would you please show me a datasheet of the used devices that uses 2,7kOhm @ 3,3V for I2C SDA & SCL .
thx.

**Reply by sfromis** [reply]
BTW. If someone wants to try two I2C busses on ESP8266, it looks like you need a build with
`#define USE_I2C_BUS2`

**Reply by HelmiBeh** [reply]
@HofeBY 
Take a look into the I2C spec of NXP at chapter 7.1. There are some diagrams to chose the pull-up resistors. For me, I use one of the smaller values to be on the fast side, if capacitance is larger. https://www.nxp.com/docs/en/user-guide/UM10204.pdf

**Reply by HofeBY2** [reply]
@HelmiBeh
Not a single of your devices is named in the pdf.
You must use PullUP's that named the same resistance to all used devices datasheet's.
OK, have fun with your chosen resitance.
